### PR TITLE
Break down main menu logic

### PR DESCRIPTION
**Pull request by fairlighteth**
# Summary

This PR breaks down the main menu logic in smaller pieces. Tries to keep the component markup dry by offloading it to helper functions.

### CODE REVIEW

File: src/custom/components/Header/mainMenu/index.tsx [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  index: number
+  url: string
+  icon?: string
+  iconSVG?: string
+  handleMobileMenuOnClick: () => void
+}
+
+// Dynamic util function to render links based on internal or external type (with or without icon image)
+function getLink({ title, externalURL, index, url, iconSVG, icon, handleMobileMenuOnClick }: LinkType) {
+  const LinkComponent = externalURL ? getLinkComponent.external : getLinkComponent.internal
```
Comment (fairlighteth) [inline src/custom/components/Header/mainMenu/index.tsx:30]: Dynamically define the component, based on if the type of link is `externalURL`

File: src/custom/components/Header/mainMenu/index.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  darkMode: boolean
+  toggleDarkMode: () => void
+  title: MAIN_MENU_TYPE['title']
+  items: MAIN_MENU_TYPE['items']
+}
+
+const getDropdown = ({ index, title, handleMobileMenuOnClick, toggleDarkMode, darkMode, items }: DropdownType) => (
+  <MenuDropdown key={index} title={title}>
+    {items &&
+      items.map(({ sectionTitle, links }, index) => {
```
Comment (alfetopito) [inline src/custom/components/Header/mainMenu/index.tsx:84]: Can `items` be falsy?
If not, `items &&` can be removed, or `items?.map` might work (haven't tested myself)

Anyway, just a nitpick

**Comment by elena-zh** [comment]
Hey @fairlighteth  could you please clarify what exactly I need to test here?

**Comment by fairlighteth** [comment]
@elena-zh I think in general that the menu component works correctly. This PR is mostly changing the logic and not functionality.

File: src/custom/components/Header/index.tsx [hunk 3]
```diff
@@ -8,40 +8,31 @@ import { useNativeCurrencyBalances } from 'state/wallet/hooks'
...
-  HeaderLinks,
   HeaderModWrapper,
   UniIcon,
-  StyledNavLink,
   AccountElement,
   BalanceText,
   HeaderControls,
   HeaderElement,
 } from './styled'
+import MainMenu from './mainMenu'
```
Comment (anxolin) [inline src/custom/components/Header/index.tsx:29]: nit: Normally the folder for components start with a capital letter

File: src/custom/components/Header/mainMenu/index.tsx (hunk 1 above)
Comment (anxolin) [reply src/custom/components/Header/mainMenu/index.tsx:30]: Why not `const LinkComponent = externalURL ? StyledNavLink : StyledNavLink` ?

File: src/custom/components/Header/mainMenu/index.tsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+function getLink({ title, externalURL, index, url, iconSVG, icon, handleMobileMenuOnClick }: LinkType) {
+  const LinkComponent = externalURL ? getLinkComponent.external : getLinkComponent.internal
+  return (
+    <LinkComponent
+      key={index}
+      exact
+      to={url}
+      href={url}
+      onClick={handleMobileMenuOnClick}
+      onClickOptional={handleMobileMenuOnClick}
```
[the remark this answers is not in the document]
Comment (fairlighteth) [reply src/custom/components/Header/mainMenu/index.tsx:38]: Yes, initially I had it only apply onClick or onClickOptional depending what component is being used. But apparently this works too. I think it simply ignores either option? `ExternalLink` does not have `onClick` only has `onClickOptional`

File: src/custom/components/Header/mainMenu/index.tsx (hunk 1 above)
Comment (fairlighteth) [reply src/custom/components/Header/mainMenu/index.tsx:30]: Yeah I guess directly referencing works and no need for the extra const.

**Comment by elena-zh** [comment]
Hey @fairlighteth , I have not noticed new critical issues besides this nitpick: I find it a bit strange when mobile menu stays expanded for some cases, for other (click and navigation) it becomes collapsed. See the video: 

https://user-images.githubusercontent.com/70885163/174630030-357ae73a-6643-433a-a01b-b86444aac753.mp4

 I think, we should remember user's state and do not collapse it automatically. 

I also have found out that 'Connect wallet' button on the Profile page  is placed differently in Ethereum and GC/Rinkeby networks (**iOS**): https://github.com/cowprotocol/cowswap/issues/669#issuecomment-1160550536

And here is the list of issues I reported when tested previous PRs, and that are still relevant for the current one: 
 - https://github.com/cowprotocol/cowswap/issues/668
 - https://github.com/cowprotocol/cowswap/issues/672
 - https://github.com/cowprotocol/cowswap/issues/276#issuecomment-1152180943
 - enhancement: https://github.com/cowprotocol/cowswap/issues/673

Thanks!

**Comment by elena-zh** [comment]
Hey @fairlighteth , I can see that the issue fixed in #653 is reproducible in the current PR. Could you please take a look at it?

**Comment by elena-zh** [comment]
@fairlighteth , thanks. I have reopened #486 issue for this.

**Pull request closed without merging**